src/reconcile/rec_write.c

```diff
@@ -598,8 +600,7 @@ __rec_init(WT_SESSION_IMPL *session, WT_REF *ref, uint32_t flags, WT_SALVAGE_COO
      * Add a write barrier to make it more likely that a thread adding an update will see this state
      * change.
      */
-    page->modify->page_state = WT_PAGE_DIRTY_FIRST;
-    WT_FULL_BARRIER();
+    atomic_store_explicit(&page->modify->page_state, WT_PAGE_DIRTY_FIRST, memory_order_release);
```

Comment:
I think a release here is not enough. We should ensure this is the first right before all the writes in the reconciliation.  The release semantic doesn't prevent the following write be reordered before it. If that happens, we can be in a race condition like this.

We enter reconciliation
We do some reconciliation work                                     Another thread does a concurrent modification
We mark the page status as dirty first

In this case, we may ignore the concurrent modification and wrongly set the page as clean after reconciliation.

I believe we should use a release atomic_thread_fence here.

Reply:
Afaict this is done under a page lock. I assumed this would protect it. Is it not the case?

The next write I see is in `__wt_txn_pinned_timestamp` which contains a barrier before the write. I don't think a barrier can be moved w.r.t to the atomic operation.

Reply:
We only have exclusive access when doing reconciliation under eviction. If it's checkpoint, it is not locked and not protected.

Yes, if `__wt_txn_pinned_timestamp` has a barrier it's OK then but I think it's dangerous to rely on that.

Reply:
I can't see that the eviction condition at `rec_write.c:77` what am I missing?

```
    /*
     * Reconciliation locks the page for two reasons:
     *    Reconciliation reads the lists of page updates, obsolete updates
     * cannot be discarded while reconciliation is in progress;
     *    In-memory splits: reconciliation of an internal page cannot handle
     * a child page splitting during the reconciliation.
     */
    WT_PAGE_LOCK(session, page);
    page_locked = true;
```

Reply:
This is in the eviction code not reconciliation code. BTW, locking the page doesn't mean exclusive access. You need to lock the page and then check the hazard pointer then you can get exclusive access. Check __wt_evict_page function if I remember correctly.

Reply:
After some more thinking, I think it is ok here to use a release store. This is because when we add an update to the update chain, we use a release compare and swap and that should prevent the write to the pages being reordered before this.